tests/lib/nested.sh

```diff
@@ -0,0 +1,210 @@
+    exit 1
+}
+
+wait_for_ssh(){
+    local service_name="$1"
+    retry=800
+    wait=1
+    while ! execute_remote true; do
+        if ! systemctl show -p ActiveState "$service_name" | grep -q "ActiveState=active"; then
+            echo "Service no longer active"
+            return 1
+        fi
+
+        retry=$(( retry - 1 ))
+        if [ $retry -le 0 ]; then
+            echo "Timed out waiting for ssh. Aborting!"
+            return 1
+        fi
+        sleep "$wait"
+    done
+}
+
+nested_wait_for_snap_command(){
+  retry=400
+  wait=1
+  while ! execute_remote command -v snap; do
+      retry=$(( retry - 1 ))
+      if [ $retry -le 0 ]; then
+          echo "Timed out waiting for snap command to be available. Aborting!"
+          exit 1
+      fi
+      sleep "$wait"
+  done
+}
+
+cleanup_nested_core_vm(){
+    # stop the VM if it is running
+    systemctl stop nested-vm-*
+
+    if [ "${ENABLE_TPM:-false}" = "true" ]; then
+        if [ -d "/tmp/qtpm" ]; then
+            rm -rf /tmp/qtpm
+        fi
+
+        # remove the swtpm
+        # TODO: we could just remove/reset the swtpm instead of removing the snap 
+        # wholesale
+        snap remove swtpm-mvo
+    fi
+
+    # delete the image file
+    rm -rf "${IMAGE_FILE}"
+}
+
+print_nested_status(){
+    SVC_NAME="nested-vm-$(systemd-escape "${SPREAD_JOB:-unknown}")"
+    systemctl status "${SVC_NAME}" || true
+    journalctl -u "${SVC_NAME}" || true
+}
+
+start_nested_core_vm_unit(){
+    # copy the image file to create a new one to use
+    # TODO: maybe create a snapshot qcow image instead?
+    mkdir -p "${WORK_DIR}"
+    cp "${SETUPDIR}/pc.img" "${IMAGE_FILE}"
+
+    # use only 2G of RAM for qemu-nested
+    if [ "${SPREAD_BACKEND}" = "google-nested" ]; then
+        PARAM_MEM="-m 4096"
+        PARAM_SMP="-smp 2"
+    elif [ "${SPREAD_BACKEND}" = "lxd-nested" ]; then
+        PARAM_MEM="-m 4096"
+        PARAM_SMP="-smp 2"
+    elif [ "${SPREAD_BACKEND}" = "qemu-nested" ]; then
+        PARAM_MEM="-m 2048"
+        PARAM_SMP="-smp 1"
+    else
+        echo "unknown spread backend ${SPREAD_BACKEND}"
+        exit 1
+    fi
+
+    PARAM_DISPLAY="-nographic"
+    PARAM_NETWORK="-net nic,model=virtio -net user,hostfwd=tcp::${SSH_PORT}-:22"
+    # TODO: do we need monitor port still?
+    PARAM_MONITOR="-monitor tcp:127.0.0.1:${MON_PORT},server,nowait"
+    PARAM_RANDOM="-object rng-random,id=rng0,filename=/dev/urandom -device virtio-rng-pci,rng=rng0"
+    PARAM_CPU=""
+    PARAM_TRACE="-d cpu_reset"
+    PARAM_LOG="-D ${WORK_DIR}/qemu.log"
+    PARAM_SERIAL="-serial file:${WORK_DIR}/serial.log"
+    PARAM_TPM=""
+
+    ATTR_KVM=""
+    if [ "$ENABLE_KVM" = "true" ]; then
+        ATTR_KVM=",accel=kvm"
+        # CPU can be defined just when kvm is enabled
+        PARAM_CPU="-cpu host"
+        # Increase the number of cpus used once the issue related to kvm and ovmf is fixed
+        # https://bugs.launchpad.net/ubuntu/+source/kvm/+bug/1872803
```

Comment:
On 2020-06-22, there was a comment asking to try again because a fix was done. Has anyone tried to follow up?

Reply:
I'll try this out and see if it crashes!